separatedunit.cpp

```diff
@@ -0,0 +1,1286 @@
+    extraQbe.length = 1;
+    qbExtra[0] = extraQbe;
+    extraQbe.cu = qubitLookup[carryIndex].cu;
+    extraQbe.start = qubitLookup[carryIndex].qb;
+    extraQbe.length = 1;
+    qbExtra[1] = extraQbe;
+    qbList.insert(qbList.end(), qbExtra.begin(), qbExtra.end());
+    OptimizeParallelBitList(qbList);
+
+    EntangleBitList(qbList);
+
+    coherentUnits[qubitLookup[start].cu]->INCSC(
+        toAdd, qubitLookup[start].qb, length, qubitLookup[overflowIndex].qb, qubitLookup[carryIndex].qb);
+}
+
+void SeparatedUnit::INCSC(bitCapInt toAdd, bitLenInt start, bitLenInt length, bitLenInt carryIndex)
+{
+    std::vector<QbListEntry> qbList(length);
+    GetParallelBitList(start, length, qbList);
+    QbListEntry carryQbe;
+    carryQbe.cu = qubitLookup[carryIndex].cu;
+    carryQbe.start = qubitLookup[carryIndex].qb;
+    carryQbe.length = 1;
+    qbList.push_back(carryQbe);
+    OptimizeParallelBitList(qbList);
+
+    EntangleBitList(qbList);
+
+    coherentUnits[qubitLookup[start].cu]->INCSC(toAdd, qubitLookup[start].qb, length, qubitLookup[carryIndex].qb);
+}
+
+void SeparatedUnit::INCBCD(bitCapInt toAdd, bitLenInt start, bitLenInt length)
+{
+    std::vector<QbListEntry> qbList(length);
+    GetParallelBitList(start, length, qbList);
+
+    EntangleBitList(qbList);
+
+    coherentUnits[qubitLookup[start].cu]->INCBCD(toAdd, qubitLookup[start].qb, length);
+}
+
+void SeparatedUnit::INCBCDC(bitCapInt toAdd, bitLenInt start, bitLenInt length, bitLenInt carryIndex)
+{
+    std::vector<QbListEntry> qbList(length);
+    GetParallelBitList(start, length, qbList);
+    QbListEntry carryQbe;
+    carryQbe.cu = qubitLookup[carryIndex].cu;
+    carryQbe.start = qubitLookup[carryIndex].qb;
+    carryQbe.length = 1;
+    qbList.push_back(carryQbe);
+    OptimizeParallelBitList(qbList);
+
+    EntangleBitList(qbList);
+
+    coherentUnits[qubitLookup[start].cu]->INCBCDC(toAdd, qubitLookup[start].qb, length, qubitLookup[carryIndex].qb);
+}
+
+void SeparatedUnit::DEC(bitCapInt toAdd, bitLenInt start, bitLenInt length)
+{
+    std::vector<QbListEntry> qbList(length);
+    GetParallelBitList(start, length, qbList);
+
+    EntangleBitList(qbList);
+
+    coherentUnits[qubitLookup[start].cu]->DEC(toAdd, qubitLookup[start].qb, length);
+}
+
+void SeparatedUnit::DECC(bitCapInt toAdd, bitLenInt start, bitLenInt length, bitLenInt carryIndex)
+{
+    std::vector<QbListEntry> qbList(length);
+    GetParallelBitList(start, length, qbList);
+    QbListEntry carryQbe;
+    carryQbe.cu = qubitLookup[carryIndex].cu;
+    carryQbe.start = qubitLookup[carryIndex].qb;
+    carryQbe.length = 1;
+    qbList.push_back(carryQbe);
+    OptimizeParallelBitList(qbList);
+
+    EntangleBitList(qbList);
+
+    coherentUnits[qubitLookup[start].cu]->DECC(toAdd, qubitLookup[start].qb, length, qubitLookup[carryIndex].qb);
+}
+
+void SeparatedUnit::DECS(bitCapInt toAdd, bitLenInt start, bitLenInt length, bitLenInt overflowIndex)
+{
+    std::vector<QbListEntry> qbList(length);
```

Comment:
(comment was placed randomly)

Scrolling through this code, it's obvious that there is a ton of duplication.  Instead of copy-and-pasting, how about providing a single function that does the entangelment boilerplate, and then takes either a reference to the memberfunction (good) or a lambda (less good but still acceptable) to call on the resulting CU?